### DISCUSSION TITLE
Initial Testing Failure

### DISCUSSION THREAD

**Discussion by GeorgeIoak** [eModbus/eModbus#215] [Q&A] [answered]
I wanted to try some initial simple tests and I'm not getting very far unfortunately.

- I have an ESP32 Dev board and external MAX485
- The output of the MAX485 is then connected to USB-to-485 Adapter and plugged into my Windows PC
- The ESP32 is flashed with the RTU04example sketch through the Arduino IDE
- I am running a MODBUS Slave simulator from https://modbustools.com/download.html
- Slave ID is set to 1, Function 03 Holding Register (4x), I know the code is set to register 0x2A but I have the simulator set to start at 0x29 for 11 registers
<img width="203" alt="image" src="https://user-images.githubusercontent.com/1467145/186797755-9cbbdc64-4339-47e8-99d1-36ef18ac0c5a.png">

Arduino serial monitor reports:
```
[0m[1;33m[E] 1140220| eModbus_RTU04example.ino.ino [  57] handleError: Error response: 01 - Illegal function code
```

The communication traffic according to the software simulator looks like there is some communication so I believe the wiring is correct but I'm not knowledgeable enough with MODBUS (yet) to begin to debug this and determine what might be the problem so any help would be greatly appreciated!

![image](https://user-images.githubusercontent.com/1467145/186797307-05d8585d-3013-4f10-9ac1-e1a8fd4f9337.png)

This is the only section that I made any changes to the example code:
```

// Definitions for this special case
#define RXPIN GPIO_NUM_34
#define TXPIN GPIO_NUM_32
#define REDEPIN GPIO_NUM_15
#define BAUDRATE 9600
#define FIRST_REGISTER 0x002A
#define NUM_VALUES 10
#define READ_INTERVAL 10000
```

**Comment by GeorgeIoak** [comment]
Figured I had to make a post to make some progress. By changing the simulator to Function 04 and increasing the quantity to 20 I was able to eliminate the error messages:
![image](https://user-images.githubusercontent.com/1467145/186799748-83d33d6f-4b9e-4e70-b6b4-77671fb0a188.png)

The problem I am now facing is that the returned data in the Arduino Serial monitor is showing all 0 yet I have changed the registers which is shown in the simulator traffic but what the example code is displaying:
```
Requested at 2670.299s:
   002A:    0.000
   002C:    0.000
   002E:    0.000
   0030:    0.000
   0032:    0.000
   0034:    0.000
   0036:    0.000
   0038:    0.000
   003A:    0.000
   003C:    0.000
```

**Comment by Miq1** [comment]
Hi,

no wonder you got a 0x01 "illegal function code" error first... The RTU04example is meant to produce requests for function code 0x04 READ_INPUT_REGISTER, hence the error if you had set up your server for 0x03 READ_HOLDING_REGISTER.

The good news is, you seem to have the wiring all correct, so communication is working in principle.

You are seeing what was responded by your server, and as I can see from the screenshot you provided the server on the PC does log the same - all zeroes. Also on the picture I see the configuration window has a remark "PLC address = 30043". The server seems to follow the convention to add 30001 to the addresses of input registers - which had been common a while ago. You will have to request from those addresses to get what you were expecting.

**Reply by GeorgeIoak** [reply]
I'll have to check the "PLC Address" and change my testing code to see if that's it but hopefully you're right so thanks for catching that!

I am confused about your statement that the PC also does log all zeros because in my 2nd screen shot at the end you can see where I entered 0, 1, 2, 3 into addresses 0x2A, 0x2B, 0x2C, 0x2D and then the last TX line is 
```
04 28 00 00 00 01 00 02 00 03 00 ...
```

So when I saw that I was wondering if perhaps it was something like a byte order difference. 

Do you happen to know of some other software that you have used for testing?

**Reply by Miq1** [reply]
Sorry I overlooked the values in the screen shot. Getting older... 🙄 

I usually take another ESP for testing or quickly will write a dedicated test client on Linux. I never used any of the available PC Modbus test applications, so I can only guess there might be a way to omit the address offsets. There is no such rule in the Modbus standard, but I know it was common to use the 30001 offset for input registers and 40001 for holding registers. There is no reason to do that; I am always wondering why someone did invent it?

**Reply by GeorgeIoak** [reply]
Getting old, don't remind me!

Just so I'm clear, for testing right now I would just have to change this line:

```
#define FIRST_REGISTER 0x002A
```

to

```
#define FIRST_REGISTER 30043
```

**Reply by Miq1** [reply]
I would say so, yes.

**Reply by GeorgeIoak** [reply]
Sorry for the screen shot but hopefully you can see it. For some reason when I changed the address this is now happening. It seems as though this software is responding differently and it's giving me an illegal address
![image](https://user-images.githubusercontent.com/1467145/186938806-f8cef97a-2d37-4be6-9267-d22a60e7592c.png)

**Reply by Miq1** [reply]
Ough, if you could do a copy & past as text next time it would be better for me.
The requests are looking good to me, ``01 04 75 5B 00 14 xx xx`` is 20 holding registers starting at 30043.

So it must be the PC program wishing something else. Interestingly it did not complain on your previous attempts with the lower addresses. There must be something else in the configuration of the test server causing this.

**Reply by GeorgeIoak** [reply]
I'm thinking that comment in the setup page of the simulator software is just a comment so I'm back to wondering if the response from the simulator software looks correct to you. For me it looks like a proper response but I have very little experience with MODBUS communication. 

Are the register values in the `Tx` line in the proper byte positions? I see that 0x2A register is returned in the 4th byte. If I enter a value of 10 in register 0x2A in the software I see the Tx line show `04 28 00 0A 00 ...`

**Reply by GeorgeIoak** [reply]
Sorry for the inconvenience, I would copy and paste values if I could but the simulation software won't let me copy data out of it's Communication Traffic Window. I did try a different software last night and using it the RTU04 example sketch also just returned all zeros in the serial monitor window. I'm hoping that this must just be something simple I'm overlooking but if I believe the communication traffic window then the software is sending a correct response but the RTU04 example sketch is not interpreting it correctly

**Reply by Miq1** [reply]
Ooops. Just found it! The responses are okay indeed, but the data you are getting is uint16_t format (single register-wise), but the sketch ``RTU04example`` is assuming there are floats coming:
```
float values[NUM_VALUES];
...
for (uint8_t i = 0; i < NUM_VALUES; ++i) {
        Serial.printf("   %04X: %8.3f\n", i * 2 + FIRST_REGISTER, values[i]);
      }
```
If you will change that to
```
uint16_t values[NUM_VALUES];
...
for (uint8_t i = 0; i < NUM_VALUES; ++i) {
        Serial.printf("   %04X: %u\n", i + FIRST_REGISTER, values[i]);
      }
```
it may look better.

**Reply by GeorgeIoak** [reply]
Great catch! I would have thought an integer would be fine being stored in a float. If I enter `10, 11, 12, 13, 14, 15` into the registers in the simulator software I get back the following in the serial monitor:
```

Requested at   20.034s:
   002A: 10
   002C: 11
   002E: 12
   0030: 13
   0032: 14
   0034: 15
   0036: 0
   0038: 0
   003A: 0
   003C: 0
----------
```

**Reply by Miq1** [reply]
Normally you were right - int fits nicely into float if assigned. But here we just have a stream of bytes, the type information is lost.

You did not yet change that ``i * 2`` to just ``i``, right? that is why the addresses are stepping by 2 still in the output.

But great you got it working.

**Reply by GeorgeIoak** [reply]
Yeah, whoops, I missed that `i * 2` to just `i` change. 

I would assume your example code works for most cases and I'm just the unlucky one who decided to test and ran across an outlying case scenario or did this catch a bug in the example code?

**Reply by Miq1** [reply]
None of both, I am afraid. The examples simply are not meant to be taken as a base for user applications, but just to show a simple, but complete case of a working code. I know most people are doing like you did and try to get the examples running with the least modifications, but then you will trip over details like it happened to you. The library has a lot of powerful features you may miss if you are avoiding writing your application from scratch. No offense intended, of course! 😉

**Reply by GeorgeIoak** [reply]
No offense taken, we're all just lazy slugs by nature! Thanks again for the hand holding and helping find out what was going wrong.

**Comment by GeorgeIoak** [comment, marked as answer]
> Ooops. Just found it! The responses are okay indeed, but the data you are getting is uint16_t format (single register-wise), but the sketch `RTU04example` is assuming there are floats coming:
> 
> ```
> float values[NUM_VALUES];
> ...
> for (uint8_t i = 0; i < NUM_VALUES; ++i) {
>         Serial.printf("   %04X: %8.3f\n", i * 2 + FIRST_REGISTER, values[i]);
>       }
> ```
> 
> If you will change that to
> 
> ```
> uint16_t values[NUM_VALUES];
> ...
> for (uint8_t i = 0; i < NUM_VALUES; ++i) {
>         Serial.printf("   %04X: %u\n", i + FIRST_REGISTER, values[i]);
>       }
> ```
> 
> it may look better.

I couldn't mark the [original reply](https://github.com/eModbus/eModbus/discussions/215#discussioncomment-3484987) as answered so hopefully this will work so the discussion can be marked as answered.
